internal/controller/dependencyupdatecheck_controller.go

```diff
@@ -80,14 +82,103 @@ func (r *DependencyUpdateCheckReconciler) getCAConfigMap(ctx context.Context) (*
 	return nil, nil
 }
 
+// Create a secret that merges all secret with the label:
+// mintmaker.appstudio.redhat.com/secret-type: registry
+// and return the new secret
+func (r *DependencyUpdateCheckReconciler) createMergedPullSecret(ctx context.Context) (*corev1.Secret, error) {
+	log := ctrllog.FromContext(ctx).WithName("DependencyUpdateCheckController")
+	ctx = ctrllog.IntoContext(ctx, log)
+
+	secretList := &corev1.SecretList{}
+	labelSelector := client.MatchingLabels{"mintmaker.appstudio.redhat.com/secret-type": "registry"}
+	listOptions := []client.ListOption{
+		client.InNamespace(MintMakerNamespaceName),
+		labelSelector,
+	}
+
+	err := r.Client.List(ctx, secretList, listOptions...)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(secretList.Items) == 0 {
+		// No secrets to merge
+		return nil, nil
+	}
+
+	log.Info(fmt.Sprintf("Found %d secrets to merge", len(secretList.Items)))
+
+	mergedAuths := make(map[string]interface{})
+	for _, secret := range secretList.Items {
+		if secret.Type == corev1.SecretTypeDockerConfigJson {
+			data, exists := secret.Data[".dockerconfigjson"]
+			if !exists {
+				// No .dockerconfigjson section
+				log.Info("Found secret without .dockerconfigjson section")
+				return nil, nil
+			}
+
+			var dockerConfig map[string]interface{}
+			if err := json.Unmarshal(data, &dockerConfig); err != nil {
+				return nil, err
+			}
+
+			auths, exists := dockerConfig["auths"].(map[string]interface{})
+			if !exists {
+				continue
+			}
+
+			for registry, creds := range auths {
+				mergedAuths[registry] = creds
+			}
+		}
+	}
+
+	mergedDockerConfig := map[string]interface{}{
+		"auths": mergedAuths,
+	}
+
+	if len(mergedAuths) == 0 {
+		log.Info("Merged auths empty, skipping creation of secret")
+		return nil, nil
+	}
+
+	mergedConfigJson, err := json.Marshal(mergedDockerConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	timestamp := time.Now().Unix()
+	name := fmt.Sprintf("renovate-image-pull-secrets-%d-%s", timestamp, RandomString(5))
+
+	newSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: MintMakerNamespaceName,
+		},
+		Type: corev1.SecretTypeDockerConfigJson,
+		Data: map[string][]byte{
+			".dockerconfigjson": []byte(mergedConfigJson),
+		},
+	}
+
+	if err := r.Client.Create(ctx, newSecret); err != nil {
+		return nil, err
+	}
+
+	return newSecret, nil
+}
+
 // createPipelineRun creates and returns a new PipelineRun
 func (r *DependencyUpdateCheckReconciler) createPipelineRun(comp component.GitComponent, ctx context.Context) (*tektonv1.PipelineRun, error) {
 
 	log := ctrllog.FromContext(ctx).WithName("DependencyUpdateCheckController")
 	ctx = ctrllog.IntoContext(ctx, log)
 	name := fmt.Sprintf("renovate-%d-%s-%s", comp.GetTimestamp(), RandomString(8), comp.GetName())
+	registry_secret, err := r.createMergedPullSecret(ctx)
```

Comment:
Currently, we obtain registry credentials from secrets prepared in mintmaker, and these credentials remain same throughout the controller's lifecycle. Since the credentials don't change, it's unnecessary to create a new secret for each PipelineRun. Instead, we can create a single secret and reuse it. For example, we can create the secret during the first reconcile, cache it, and avoid recreating it in subsequent reconciles. This cached secret can then be mounted to all PipelineRuns.

It was not a big issue we create one secret for each job, because we only have 1-5 jobs in each round of Renovate run, but now we have thousands of PipelineRuns in each round.

Reply:
I had this doubt. Thanks for raising the concern. Let me fix that.